envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt

```diff
@@ -32,12 +34,45 @@ class SnapshotUpdater(
     )
 
     fun start(changes: Flux<List<LocalityAwareServicesState>>): Flux<List<LocalityAwareServicesState>> {
-        // see GroupChangeWatcher
         return Flux.combineLatest(
-            onGroupAdded,
-            changes,
+            groups(),
+            services(changes),
```

Comment:
Unfortunately I think we create a race condition between `groups()` and `services(changes)`, because both call `updateSnapshotForGroup()`. Call from `groups()` may win the race and we will end up with snapshot for a group that is based on stale global snapshot.

I think we can fix the problem following way:
* Create some wrapper for results: `data class Wrapper(val changedGroups: List<Group>? = null, val xdsSnapshot: Snapshot? = null, val adsSnapshot: Snapshot? = null)`
* change combineLatest() to merge()
* replace `groups()` with simply `onGroupAdded` flux
* rename `services(changes)` to `snapshots(changes)` which should only calculate global snapshots and return all groups
* then call `doOnNext()` on the merged flux and call `updateSnapshotForGroup()` for each group from the flux

Probably I am missing something but that's just an idea to consider.